# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/webrtc/jsapi/PeerConnectionCtx.cpp
# Commit: 786ea95e928b
# Full Hash: 786ea95e928b7bcec53a56b50f1d0ff136c76a62
# Author: Andreas Pehrson <apehrson@mozilla.com>
# Date: 2021-11-01 18:17:33
# Regressor Bug: 1654112
# File Overlap Count: 1
# Description:
#   Bug 1654112 - Move NoTrialsConfig into SharedWebrtcState and make WebrtcCallWrapper keep a ref to it. r=ng
#   
#   SharedWebrtcState when managed by PeerConnectionCtx is destroyed on main thread.
#   With interactions with call instances being async, there is no guarantee that
#   the objects owned by SharedWebrtcState (and referenced through raw pointers by
# ==============================================================================

diff -r 2a373d864092 -r 786ea95e928b dom/media/webrtc/jsapi/PeerConnectionCtx.cpp
--- a/dom/media/webrtc/jsapi/PeerConnectionCtx.cpp	Wed Apr 14 17:40:24 2021 +0200
+++ b/dom/media/webrtc/jsapi/PeerConnectionCtx.cpp	Fri Apr 16 14:44:09 2021 +0200
@@ -171,10 +171,12 @@
 SharedWebrtcState::SharedWebrtcState(
     RefPtr<AbstractThread> aCallWorkerThread,
     webrtc::AudioState::Config&& aAudioStateConfig,
-    RefPtr<webrtc::AudioDecoderFactory> aAudioDecoderFactory)
+    RefPtr<webrtc::AudioDecoderFactory> aAudioDecoderFactory,
+    UniquePtr<webrtc::WebRtcKeyValueConfig> aTrials)
     : mCallWorkerThread(std::move(aCallWorkerThread)),
       mAudioStateConfig(std::move(aAudioStateConfig)),
-      mAudioDecoderFactory(std::move(aAudioDecoderFactory)) {}
+      mAudioDecoderFactory(std::move(aAudioDecoderFactory)),
+      mTrials(std::move(aTrials)) {}
 
 SharedWebrtcState::~SharedWebrtcState() = default;
 
@@ -479,14 +481,14 @@
                                     webrtc::TaskQueueFactory::Priority::NORMAL)
             .release());
 
+    UniquePtr<webrtc::WebRtcKeyValueConfig> trials =
+        WrapUnique(new NoTrialsConfig());
+
     mSharedWebrtcState = MakeAndAddRef<SharedWebrtcState>(
         new CallWorkerThread(std::move(callWorkerThread)),
         std::move(audioStateConfig),
-        already_AddRefed(CreateBuiltinAudioDecoderFactory().release()));
-
-    if (!mTrials) {
-      mTrials.reset(new NoTrialsConfig());
-    }
+        already_AddRefed(CreateBuiltinAudioDecoderFactory().release()),
+        std::move(trials));
   }
   mPeerConnections[aKey] = aPeerConnection;
 }